torch_np/_dtypes_impl.py

```diff
@@ -49,3 +49,82 @@ def result_type_impl(*tensors):
         dtyp = _cd._result_type_dict[dtyp][curr.dtype]
 
     return dtyp
+
+
+# ### NEP 50 helpers ###
+
+SCALAR_TYPES = (int, bool, float, complex)
+
+
+def _dtype_for_scalar(py_type):
+    return {
+        bool: torch.bool,
+        int: torch.int64,
+        float: torch.float64,
+        complex: torch.complex128,
+    }[py_type]
+
+
+categories = [
+    (torch.bool,),
+    (torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64),
+    (torch.float16, torch.float32, torch.float64),
+    (torch.complex64, torch.complex128),
+]
+
+
+def category(dtyp):
+    for j, cat in enumerate(categories):
+        if dtyp in cat:
+            return j
+    raise ValueError(f"unknown dtype {dtyp}")
+
+
+dtype_for_cat = {0: torch.bool, 1: torch.int64, 2: torch.float64, 3: torch.complex128}
+
+
+def nep50_to_tensors(x1, x2):
+    """If either of inputs is a python scalar, type-promote with NEP 50.
+
+    NB: NEP 50 mandates RuntimeWarnings on some overflows. We do not emit them:
+    we either raise an OverflowError or silently do the computation.
```

Comment:
The main reason I "mandate" it, is to point out that users will get warnings on some of the most surprising changes (integer overflows due to a now smaller dtype).

Reply:
Agreed, this makes perfect sense for NumPy. At some point I guess these can become silent in NumPy, but this is long time away anyway.